File: apps/android/app/src/main/java/page/stephens/cascade/settings/SettingsStore.kt
package page.stephens.cascade.settings

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.map

private val Context.dataStore by preferencesDataStore("cascade-settings")

/**
 * Stores the cascade-core settings blob (JSON produced by the Rust
 * `PersistSettings` effect). The JSON shape is opaque to Kotlin — we just
 * round-trip it.
 */
class SettingsStore(private val context: Context) {
    private val key = stringPreferencesKey("settings_v1")
    // Lifetime listening ledger — a separate blob from settings, so the two
    // evolve and fail independently (mirrors `cascade.listening.v1` on web).
    private val listeningKey = stringPreferencesKey("listening_v1")

    suspend fun read(): String? =
        context.dataStore.data.map { it[key] }.first()

    suspend fun write(json: String) {
        context.dataStore.edit { it[key] = json }
    }

    suspend fun readListening(): String? =
        context.dataStore.data.map { it[listeningKey] }.first()

    suspend fun writeListening(json: String) {
        context.dataStore.edit { it[listeningKey] = json }
    }
}

File: apps/apple/CascadeShared/Persistence/SettingsStore.swift
import Foundation

/// Stores the cascade-core settings blob (JSON produced by `Effect.persistSettings`)
/// as `~/Library/Application Support/Cascade/settings.json`.
///
/// The JSON shape is opaque to Swift — the Rust core owns the schema. We just
/// round-trip the bytes.
final class SettingsStore {
    private let url: URL
    /// Lifetime listening ledger — a separate file from settings, so the two
    /// evolve and fail independently (mirrors cascade.listening.v1 on web).
    private let listeningUrl: URL

    init() {
        let fm = FileManager.default
        let support = (try? fm.url(for: .applicationSupportDirectory,
                                   in: .userDomainMask,
                                   appropriateFor: nil,
                                   create: true))
            ?? URL(fileURLWithPath: NSHomeDirectory()).appendingPathComponent("Library/Application Support")
        let dir = support.appendingPathComponent("Cascade", isDirectory: true)
        try? fm.createDirectory(at: dir, withIntermediateDirectories: true)
        self.url = dir.appendingPathComponent("settings.json")
        self.listeningUrl = dir.appendingPathComponent("listening.json")
    }

    /// Returns nil if the file doesn't exist or can't be read — the bridge
    /// then boots with defaults.
    func readSafely() -> String? { Self.read(url) }

    func writeSafely(_ json: String) { Self.write(json, to: url, label: "settings") }

    func readListeningSafely() -> String? { Self.read(listeningUrl) }

    func writeListeningSafely(_ json: String) { Self.write(json, to: listeningUrl, label: "listening") }

    private static func read(_ url: URL) -> String? {
        guard let data = try? Data(contentsOf: url) else { return nil }
        return String(data: data, encoding: .utf8)
    }

    private static func write(_ json: String, to url: URL, label: String) {
        guard let data = json.data(using: .utf8) else { return }
        do {
            try data.write(to: url, options: [.atomic])
        } catch {
            NSLog("[Cascade] \(label) write failed: \(error)")
        }
    }

    /// For the "Reveal in Finder" debug action.
    var fileURL: URL { url }
}
